CORE-4439 Gateway certificate validation

williamvigorr3 opened the pull request:
Move the session certificate revocation check from the link manager to the gateway. Communication between the LinkManager and Gateway is done via the message bus. Using a RPCSender: RevocationCheckerClient in the LinkManager, as well as a RPCResponderProcessor: RevocationCheckProcessor in the Gateway. If in the future other components also want to delegate the revocation check to the Gateway, then we should pull out the RevocationCheckerClient into a common module. During session negotiation the Link Manager will make a request to the gateway.

Merge (https://github.com/corda/corda-api/pull/683) first.

williamvigorr3 committed:
CORE-4439 Add component to Gateway for cert validation

Use an RPC processor

williamvigorr3 committed:
Add revocation checker to the Gateway which can be triggered by an RPC sender

williamvigorr3 committed:
Add unit test of RevocationChecker

williamvigorr3 committed:
Use client to check with gateway for revocation

williamvigorr3 committed:
Fix broken test

williamvigorr3 committed:
Use API build

williamvigorr3 committed:
Tidy code

williamvigorr3 committed:
Update API build

williamvigorr3 committed:
Merge remote-tracking branch 'origin/release/os/5.0' into WillV/CORE-4439-Gateway-Cert-Validation

williamvigorr3 committed:
Fix detekt

williamvigorr3 committed:
Export revocation checker package

williamvigorr3 committed:
Improve unit testing

williamvigorr3 committed:
Increase timeout in RevocationCheckerTest

williamvigorr3 committed:
Fix detekt

williamvigorr3 committed:
Merge remote-tracking branch 'origin/release/os/5.0' into WillV/CORE-4439-Gateway-Cert-Validation

williamvigorr3 committed:
Move package of avro messages

williamvigorr3 committed:
Fix tests

williamvigorr3 committed:
Fix broken test

williamvigorr3 committed:
Increase timeout

williamvigorr3 committed:
Close subscription on every test case

williamvigorr3 committed:
Minor changes to CertificateValidatorTest

williamvigorr3 committed:
Revert accidentally commited file

dimosr commented on components/gateway/src/main/kotlin/net/corda/p2p/gateway/certificates/RevocationChecker.kt:
```diff
@@ -0,0 +1,141 @@
+            val revocationChecker = when (revocationConfig.mode) {
+                RevocationConfigMode.OFF -> AllowAllRevocationChecker
+                RevocationConfigMode.SOFT_FAIL, RevocationConfigMode.HARD_FAIL -> {
+                    val certPathBuilder = CertPathBuilder.getInstance("PKIX")
```
```suggestion
                    val certPathBuilder = CertPathBuilder.getInstance(certificateAlgorithm)
```

dimosr commented on components/gateway/src/main/kotlin/net/corda/p2p/gateway/certificates/RevocationChecker.kt:
```diff
@@ -0,0 +1,141 @@
+                )
+                return
+            }
+            val pkixRevocationChecker = getCertCheckingParameters(trustStore, revocationMode.toRevocationConfig())
```
```suggestion
            val pkixParams = getCertCheckingParameters(trustStore, revocationMode.toRevocationConfig())
```
nit

dimosr commented on components/gateway/src/main/kotlin/net/corda/p2p/gateway/certificates/RevocationChecker.kt:
```diff
@@ -0,0 +1,141 @@
+            try {
+                certPathValidator.validate(certificateChain, pkixRevocationChecker)
+            } catch (exception: CertPathValidatorException) {
+                respFuture.complete(RevocationCheckStatus.REVOKED)
```
I think when revocation fails we get a quite human-friendly error message on the reason, right?

Should we add an extra field in the response to include that?

williamvigorr3 replied:
I did this and reworked the response.

dimosr commented on components/gateway/src/main/kotlin/net/corda/p2p/gateway/certificates/RevocationChecker.kt:
```diff
@@ -0,0 +1,141 @@
+            respFuture.complete(RevocationCheckStatus.ACTIVE)
+        }
+
+        private fun RevocationMode.toRevocationConfig(): RevocationConfig {
```
should we change the `getCertCheckingParameters` to take `RevocationConfigMode` as parameter?

If we did that, I think we wouldn't need this.

williamvigorr3 replied:
We also use `getCertCheckingParameters` in the http client. So it takes the more extensive `RevocationConfigMode` which also includes the mode `OFF`. My opinion (weakly) favours leaving the code as is and so we don't have to duplicate some of the logic inside `getCertCheckingParameters`.

dimosr replied:
ah yes, I forgot that we have two different modes now. So, when commenting, I was thinking we could just do `revocationCheck.mode` at the invocation site in `HttpClient`. But, we'd have to do the conversions between the two types anyway.

dimosr commented on libs/crypto/crypto-utils/src/main/kotlin/net/corda/crypto/utils/CertificateUtils.kt:
```diff
@@ -146,4 +146,9 @@ object AllowAllRevocationChecker : PKIXRevocationChecker() {
\ No newline at end of file
+}
+
+data class KeyStoreWithPem(
+    val keyStore: KeyStore,
+    val pemKeyStore: List<PemCertificate>
+)
```
These represent the same certificates, right? Is there a reason for passing them around together vs passing only one of the two formats and converting between them, where needed?

williamvigorr3 replied:
Two reasons: Firstly we will fail earlier if the trust store is invalid (when the group gets updated) (perhaps this is unlikely to happen because we hopefully validate this when the group policy is uploaded). Secondly if we store both formats we don't have to convert between the two (in the LinkManager) every time we do a revocation check.

dimosr replied:
yes, I think we are safe on the first point. On the second point, won't the session manager end up performing a lookup every time, which internally performs a conversion anyway? Or do we use both, in which case we'd need extra conversions?

I am just wondering whether we are prematurely optimising here, because this might end up being a considerable amount of memory which might end up having adverse effects due to GC churn. So, in the absence of concrete data doing the simpler thing might be preferable.

williamvigorr3 committed:
Wrap avro enum

dimosr commented on libs/p2p-crypto/src/main/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidator.kt:
```diff
@@ -74,23 +76,21 @@ class CertificateValidator(
+            RevocationCheckMode.HARD_FAIL -> RevocationMode.HARD_FAIL
+            RevocationCheckMode.SOFT_FAIL -> RevocationMode.SOFT_FAIL
+        }
+        val revocationStatus = checkRevocation(RevocationCheckRequest(pemCertificates, trustStore, revocationMode))
```
should we skip checking if the mode is `OFF`?

williamvigorr3 replied:
We are doing that (return on line 81).

dimosr replied:
d'uh, my brain must have short-circuited the `when` clause waiting for an if-clause right after it.

williamvigorr3 committed:
Fix tests

dimosr commented on components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/RevocationCheckerClient.kt:
```diff
@@ -0,0 +1,49 @@
+            rpcSender.sendRequest(request).getOrThrow(timeout)
+        } catch (except: TimeoutException) {
+            logger.warn("Revocation request sent to the gateway timed out.")
+            return RevocationCheckStatus.REVOKED
```
I believe `SOFT_FAIL` mode allows one to consider a certificate non-expired when revocation status check was impossible due to network errors.

Should we adjust the response accordingly here in that case?

And also, see my other comment about not sending requests when mode is `OFF`. If we send them, I think it's possible for them to timeout and then we return back `REVOKED` here, which wouldn't be ideal.

williamvigorr3 replied:
I think if this request times out it means the link manager can't talk to the gateway through the message bus. If this is the case it is unlikely to be a situation we can recover from quickly even if the certificate is treated as ACTIVE.

dimosr replied:
I think it doesn't necessarily mean that the link manager can't talk to the gateway at all. It is possible this request is not completed even though the link manager can transfer data to the gateway through other means fine (e.g. via the `link.out` topic), e.g. because there was some rebalance and the request got lost. In general, rpc request/reply pattern is expected to be unreliable.

dimosr reviewed (commented):
I did a first pass, leaving a few comments. And I'll come back to have a quick look at tests.

dimosr commented on components/gateway/src/integrationTest/kotlin/net/corda/p2p/gateway/certificates/RevocationCheckerTest.kt:
```diff
@@ -0,0 +1,89 @@
+    private companion object {
+        val futureTimeOut: Duration = Duration.ofSeconds(120)
+        val lifecycleCoordinatorFactory = LifecycleCoordinatorFactoryImpl(LifecycleRegistryImpl(), LifecycleCoordinatorSchedulerFactoryImpl())
+        val topicService = TopicServiceImpl()
```
I think we should avoid adding more dependencies on the (deprecated) in-memory version, especially given we'll be removing it very soon.

dimosr replied:
One more think about this before I go on PTO.. I was thinking that it might be better to integration test the message bus as part of the message bus library once, instead of every component integration testing it.

dimosr commented on components/gateway/src/test/kotlin/net/corda/p2p/gateway/messaging/certificates/RevocationCheckerTest.kt:
```diff
@@ -0,0 +1,67 @@
+        @Suppress("UNCHECKED_CAST")
+        (context.arguments()[1] as  () -> RPCSubscription<RevocationCheckRequest, RevocationCheckStatus>)()
+    }
+    private val revocationChecker = RevocationChecker(subscriptionFactory, mock(), mock())
```
I think this is unused? And if you remove it, there are a couple of other things that are unused and can be removed e.g. the subscription.

williamvigorr3 replied:
Only the `val` is unused. I moved it into an init block.

dimosr commented on libs/p2p-crypto/src/test/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidatorTest.kt:
```diff
@@ -46,15 +51,19 @@ class CertificateValidatorTest {
     @Test
     fun `certificate fails validation if X500 name doesn't match`() {
-        val validator = CertificateValidator(RevocationCheckMode.HARD_FAIL, trustStore, certPathValidator, certificateFactory)
+        val validator = CertificateValidator(
```
should we now add a unit test for the scenario where revocation checks are enabled and the check returns negative (certificate is revoked)?

And _maybe_ one when the mode is OFF, where regardless of whether it's revoked validation succeeds (because no check is performed)?

williamvigorr3 replied:
We integration test these scenarios at the moment. We probably can unit test them instead, to do that we would probably need to mock the construction of PKIXBuilderParameters. Do you think it is worth it?

dimosr replied:
~We could also maintain the integration test nature of it, but switch to using the actual in-memory/DB version of the library.~ scratch that, I thought your response was to another comment. If it's covered by integration tests, it should be ok.

williamvigorr3 replied:
This is tested in the other `CertificateValidatorTest`.

williamvigorr3 committed:
Add reason to response when checking for revocation in gateway

Also if communication with message bus and SOFT validation is on
return unrevoked.

williamvigorr3 committed:
Only pass certificate around as PEM string

williamvigorr3 committed:
Improve error handling inside CertificateValidator + add unit tests

williamvigorr3 committed:
Remove message bus from RevocationCheckerTest

williamvigorr3 committed:
Merge remote-tracking branch 'origin/release/os/5.0' into WillV/CORE-4439-Gateway-Cert-Validation

williamvigorr3 committed:
Update corda api build

dimosr commented on libs/p2p-crypto/src/main/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidator.kt:
```diff
@@ -83,8 +84,17 @@ class CertificateValidator(
+            }?. let { specificCert ->
+                throw InvalidPeerCertificate("The certificate failed a revocation check (in the gateway): " + status.reason, specificCert)
+            }
+            throw InvalidPeerCertificate("The certificate failed a revocation check (in the gateway): " + status.reason, x509CertChain)
```
In theory, this library is supposed to be agnostic to higher-levels. So, it might be better to remove the `(in the gateway)` bit from the  message.

dimosr commented on libs/p2p-crypto/src/main/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidator.kt:
```diff
@@ -34,11 +34,15 @@ class CertificateValidator(
+                }
+            })
+        } catch (except: CertificateException) {
+            throw InvalidPeerCertificate("Error passing the certificate from a pem file: \n" + except.message)
```
Did you mean `parsing` (instead of `passing` here)?

williamvigorr3 replied:
🤦

dimosr commented on libs/p2p-crypto/src/main/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidator.kt:
```diff
@@ -74,23 +81,32 @@ class CertificateValidator(
+        val trustStore = convertToKeyStore(certificateFactory, pemTrustStore, "session")
+            ?: throw InvalidPeerCertificate("Could not validate certificate, the trust store could not be read from a pem file.")
+        val pkixParams = PKIXBuilderParameters(trustStore, X509CertSelector())
+        pkixParams.addCertPathChecker(AllowAllRevocationChecker)
```
can we add a one-line comment here to assist people understand why we do this, e.g.
```
// revocation checks are performed separately (see checkRevocation() function), here we just validate the certificate chain. 
```

dimosr reviewed (commented):
LGTM - just a few very minor comments

williamvigorr3 committed:
Respond to review comments

williamvigorr3 committed:
Merge remote-tracking branch 'origin/release/os/5.0' into WillV/CORE-4439-Gateway-Cert-Validation

williamvigorr3 committed:
Update api version

williamvigorr3 committed:
Merge remote-tracking branch 'origin/release/os/5.0' into WillV/CORE-4439-Gateway-Cert-Validation

williamvigorr3 committed:
Use corda api from release/os/5.0
